app/src/androidTest/java/org/mozilla/focus/activity/helpers/HostScreencapScreenshotStrategy.java

```diff
@@ -0,0 +1,107 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.focus.activity.helpers;
+
+import android.content.Context;
+import android.graphics.Bitmap;
+import android.graphics.BitmapFactory;
+import android.graphics.Canvas;
+import android.graphics.Color;
+import android.support.test.InstrumentationRegistry;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.ConnectException;
+import java.net.HttpURLConnection;
+import java.net.SocketTimeoutException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+
+import tools.fastlane.screengrab.ScreenshotCallback;
+import tools.fastlane.screengrab.ScreenshotStrategy;
+
+/**
+ * This ScreenshotStrategy implementation pings the host system (the computer the emulator is
+ * running on) to take a screenshot using "screencap" via adb. After that the screenshot is read
+ * from the internal app storage and passed back to fastlane/screengrab using the provided callback.
+ */
+public class HostScreencapScreenshotStrategy implements ScreenshotStrategy {
+    private static final int CONNECT_TIMEOUT = 500;
+    private static final int READ_TIMEOUT = 2000;
+    private static final String HOST_LOOPBACK = "10.0.2.2";
+    private static final int PORT = 9771;
+
+    @Override
+    public void takeScreenshot(String screenshotName, ScreenshotCallback screenshotCallback) {
+        takeScreenshotViaHost();
+
+        Bitmap bitmap = readScreenshotFromStorage();
+
+        if (bitmap == null) {
+            bitmap = createDummyScreenShot();
+        }
+
+        screenshotCallback.screenshotCaptured(screenshotName, bitmap);
+    }
+
+    private void takeScreenshotViaHost() {
+        try {
+            final HttpURLConnection connection = (HttpURLConnection) new URL("http://" + HOST_LOOPBACK + ":" + PORT).openConnection();
```

Comment:
This is awesome, just curious, would it make any more reliable if we define a connection at the beginning of the test in an initializer, and close the connection when the object is cleared?  It's just a suggestion

Reply:
Locally it's 100% reliable. So I guess that in automation we just hit the timeouts. But re-using an existing connection might make it faster. :)